src/hotspot/share/gc/g1/g1BlockOffsetTable.inline.hpp

```diff
@@ -141,9 +135,16 @@ inline HeapWord* G1BlockOffsetTablePart::block_at_or_preceding(const void* addr)
   return q;
 }
 
-inline HeapWord* G1BlockOffsetTablePart::forward_to_block_containing_addr_const(HeapWord* q, HeapWord* n,
-                                                                                const void* addr) const {
+inline HeapWord* G1BlockOffsetTablePart::forward_to_block_containing_addr(HeapWord* q, HeapWord* n,
+                                                                          const void* addr) const {
   while (n <= addr) {
+    // When addr is not covered by the block starting at q we need to
+    // step forward until we find the correct block. With the BOT
+    // being precise, we should never have to step through more than
+    // a single card.
+    assert(_bot->index_for(n) == _bot->index_for(addr),
+           "BOT not precise. Index for n: " SIZE_FORMAT " must be equal to the index for addr: " SIZE_FORMAT,
+           _bot->index_for(n), _bot->index_for(addr));
```

Comment:
I think this method is unnecessary, and always returns `q`. In `block_start()`, the only caller afaics, we already made sure that `addr` is < `top()`, i.e. we always ever ask for addresses where the BOT is complete.
For a complete BOT, 
```
  HeapWord* q = block_at_or_preceding(addr);
  HeapWord* n = q + block_size(q);
```
`q` is necessarily the closest start of the block, and `n` always > `addr`.
It would be different if `block_start` allowed queries after `top`, then we would need to skip at most one block as mentioned in the comment, but then I'd just write:
``` if (addr >= top()) { return _top; }```

Unless I'm missing something crucial here?

Reply:
You're missing the case where `addr` is not covered by the first object ("block") in the card. So when asking for `block_at_or_preciding(addr)` we get an object that is stretching into the card same card as `addr` map to, but it is not certain that it is the object containing `addr`.

Reply:
My initial thought was also that this should not be needed, but the code proved me wrong. It might be possible to refactor the code even further to avoid this case.

Reply:
Okay, understood. I have been too fixated on the case with remembered sets to disregard other cases, looking at the code I even found interesting ones :)
```
           card1       card2
   -----+----------+----------+---------
       A|AAABBBCCCD|DDD       |
   -----+----------+----------+---------
       ^    ^  ^
       q    n  addr
```   
Something like this: we get asked by `addr` somewhere in card card1, `block_preceeding_addr(addr)` is `q` and `n` is the end of the block.

Agree, then the code (and the comment) is correct.

Maybe it is really worth to optimize the card scanning case.

Reply:
Yes, if we can avoid doing the call to `block_start()` when we know we have a valid object start that could be a nice optimization.